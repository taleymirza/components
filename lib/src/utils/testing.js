import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(mount(component)).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    const dest = computed[key] || (computed[key] = {})
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return div.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}
