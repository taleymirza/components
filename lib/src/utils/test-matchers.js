import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toSetDefaultTheme(Component) {
    const wrapper = mount(<Component />)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  }
})
