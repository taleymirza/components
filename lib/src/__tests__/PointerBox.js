import React from 'react'
import {PointerBox} from '..'
import {render} from '../utils/testing'

describe('PointerBox', () => {
  it('renders a <Caret> in <BorderBox> with relative positioning', () => {
    expect(render(<PointerBox />)).toMatchSnapshot()
  })

  it('passes the "borderColor" prop to both <BorderBox> and <Caret>', () => {
    expect(render(<PointerBox borderColor="red.5" />)).toMatchSnapshot()
  })

  it('passes the "bg" prop to both <BorderBox> and <Caret>', () => {
    expect(render(<PointerBox bg="red.5" />)).toMatchSnapshot()
  })

  it('has default theme', () => {
    expect(PointerBox).toSetDefaultTheme()
  })

  it('respects the "as" prop', () => {
    expect(render(<PointerBox as="span" />).type).toEqual('span')
  })
})
