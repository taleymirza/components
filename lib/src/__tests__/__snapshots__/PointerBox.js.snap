// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointerBox passes the "bg" prop to both <BorderBox> and <Caret> 1`] = `
.c1 {
  background-color: #d73a49;
  border: 1px solid;
  border-color: #e1e4e8;
  border-radius: 3px;
}

.c0 {
  position: relative;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <svg
      height={16}
      style={
        Object {
          "left": "50%",
          "marginLeft": -8,
          "marginTop": "-1px",
          "pointerEvents": "none",
          "position": "absolute",
          "top": "100%",
        }
      }
      width={16}
    >
      <g
        transform="translate(8,0)"
      >
        <path
          d="M-8,0L0,8L8,0L-8,0Z"
          fill="#d73a49"
        />
        <path
          d="M-8,0L0,8L8,0"
          fill="none"
          stroke="#e1e4e8"
          strokeWidth={1}
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`PointerBox passes the "borderColor" prop to both <BorderBox> and <Caret> 1`] = `
.c1 {
  border-color: #d73a49;
  border: 1px solid;
  border-radius: 3px;
}

.c0 {
  position: relative;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <svg
      height={16}
      style={
        Object {
          "left": "50%",
          "marginLeft": -8,
          "marginTop": "-1px",
          "pointerEvents": "none",
          "position": "absolute",
          "top": "100%",
        }
      }
      width={16}
    >
      <g
        transform="translate(8,0)"
      >
        <path
          d="M-8,0L0,8L8,0L-8,0Z"
          fill="#fff"
        />
        <path
          d="M-8,0L0,8L8,0"
          fill="none"
          stroke="#d73a49"
          strokeWidth={1}
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`PointerBox renders a <Caret> in <BorderBox> with relative positioning 1`] = `
.c1 {
  border: 1px solid;
  border-color: #e1e4e8;
  border-radius: 3px;
}

.c0 {
  position: relative;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <svg
      height={16}
      style={
        Object {
          "left": "50%",
          "marginLeft": -8,
          "marginTop": "-1px",
          "pointerEvents": "none",
          "position": "absolute",
          "top": "100%",
        }
      }
      width={16}
    >
      <g
        transform="translate(8,0)"
      >
        <path
          d="M-8,0L0,8L8,0L-8,0Z"
          fill="#fff"
        />
        <path
          d="M-8,0L0,8L8,0"
          fill="none"
          stroke="#e1e4e8"
          strokeWidth={1}
        />
      </g>
    </svg>
  </div>
</div>
`;
